crates/bevy_ecs/src/schedule/parallel_executor.rs

```diff
@@ -205,74 +199,122 @@ impl ExecutorStage {
+                log::trace!(
+                    "prepare {} {} with {} dependents and {} dependencies",
+                    system_index,
+                    system.lock().name(),
+                    self.system_dependents[system_index].len(),
+                    self.system_dependencies[system_index].count_ones(..)
+                );
 
-            // if all system dependencies are finished, queue up the system to run
-            if self.system_dependencies[system_index].is_subset(&self.finished_systems) {
-                let system = systems[system_index].clone();
+                for dependency in self.system_dependencies[system_index].ones() {
+                    log::trace!("  * Depends on {}", systems[dependency].lock().name());
+                }
 
-                // handle thread local system
-                {
-                    let system = system.lock();
-                    if let ThreadLocalExecution::Immediate = system.thread_local_execution() {
-                        if systems_currently_running {
-                            // if systems are currently running, we can't run this thread local system yet
-                            continue;
-                        } else {
-                            // if no systems are running, return this thread local system to be run exclusively
-                            return RunReadyResult::ThreadLocalReady(system_index);
+                // This event will be awaited, preventing the task from starting until all
+                // our dependencies finish running
+                let ready_event = self.ready_events[system_index].clone();
+
+                // Clear any dependencies on systems before this range of systems. We know at this
+                // point everything before start_system_index is finished, and our ready_event did
+                // not exist to be decremented until we started processing this range
+                if start_system_index != 0 {
+                    if let Some(ready_event) = ready_event.as_ref() {
+                        for dependency in self.system_dependencies[system_index].ones() {
+                            log::trace!("  * Depends on {}", dependency);
+                            if dependency < start_system_index {
+                                ready_event.decrement();
+                            }
                         }
                     }
                 }
 
-                // handle multi-threaded system
-                let sender = self.sender.clone();
-                self.running_systems.insert(system_index);
+                let world_ref = &*world;
+                let resources_ref = &*resources;
+
+                // For every task that depends on this one, we need to decrement their ready event.
+                // Gather a list of those here to pass into the task. They will be decremented when
+                // the task finishes
+                let mut trigger_events = Vec::new();
```

Comment:
I don't think the upstream crate event-listener supports resetting, so I don't know that we can get this completely allocation free or that CountdownEvent.reset(n) will be pretty but it's probably worth at least trying it to see what it would look like.

Reply:
On second thought, it looks like event automatically resets if all listeners are notified so this may not be too bad